rep-2004.rst

```diff
@@ -0,0 +1,808 @@
+   iii. May have a "quality declaration" document, which declares the quality level and justifies how the package meets each of the requirements
+
+        a. *Must have a section in the repository's ``README`` which contains the "quality declaration" or links to it*
+        b. *May register with a centralized list of 'Level 3' packages, if one exists, to allow for peer review of the claim*
+
+4. **Testing:**
+
+   i. No explicit testing requirements, though covering some if not all of the features with tests is recommended
+
+5. **Dependencies:**
+
+   i. May have direct runtime "ROS" dependencies which are not 'Level 3' dependencies, but they should still be documented in the quality declaration
+
+6. **Platform Support:**
+
+   i. Must support all target platforms for the package's ecosystem.
+
+      * For ROS 2 this means supporting all tier 1 platforms, as defined in `REP-2000 <https://www.ros.org/reps/rep-2000.html#support-tiers>`_
+
+If the above points are satisfied then a package can be considered 'Level 3'.
+Refer to the detailed description of the requirements following the Quality Level 1 section above for more information.
+
+Quality Level 4
+^^^^^^^^^^^^^^^
+
+These are demos, tutorials, or experiments.
+They don't have strict requirements, but are not excluded from having good documentation or tests.
+For example, this might be a tutorial package which is not intended for reuse but has excellent documentation because it serves primarily as an example to others.
+
+Package Requirements
+~~~~~~~~~~~~~~~~~~~~
+
+Requirements to be considered a 'Level 4' package:
+
+1. **Version Policy:**
+
+   i. No requirements, but having a policy is still recommended (e.g. ``semver``), even if the version is not yet stable (e.g. >= 1.0.0 for ``semver``)
+
+2. **Change Control Process:**
+
+   i. No explicit change control process required, but still recommended
+
+3. **Documentation:**
+
+   i. Must have a declared license or set of licenses
+   ii. Must have a copyright statement in each source file
+
+4. **Testing:**
+
+   i. No explicit testing requirements, though covering some if not all of the features with tests is recommended
+
+5. **Dependencies:**
+
+   i. No restrictions
+
+6. **Platform Support:**
+
+   i. Must support all target platforms for the package's ecosystem.
+
+      * For ROS 2 this means supporting all tier 1 platforms, as defined in `REP-2000 <https://www.ros.org/reps/rep-2000.html#support-tiers>`_
+
+Any package that does not claim to be 'Level 3' or higher is automatically 'Level 4'.
+Refer to the detailed description of the requirements following the Quality Level 1 section above for more information.
+
+Quality Level 5
+^^^^^^^^^^^^^^^
+
+Packages in this category cannot even meet the simple 'Level 4' requirements, and for that reason should not be used.
+'Level 5' has no requirements, but we still strongly recommend all packages have a license and copyright statement, ideally in each file.
```

Comment:
I would like to propose that quality level 5 also requires to have a license in the repo (either a license.txt or a field in the package.xml) because without one, there's no safe / legal way that I am aware of to use that code. I think even the lowest quality should require some explicit licensing somewhere in the project.

Reply:
When I lobbied to get Level 5 included it was intended to be the catch all for everything that doesn’t adhere to QL4. 

I explicitly wanted a level that we could use in a future central quality repository that meant “this package is actually legally or otherwise dangerous to use”.

Reply:
Huh interesting. Ok, I can get behind that. I wonder though if someone hasn't bothered with a license or any minimal level to get to L4 if they'll ever think to declare it at all to be indexed. But I do like having that option available for other people to tag it as L5 (e.g. folks poking to it and saying "that's dangerous") even if they are unaware.

Reply:
> When I lobbied to get Level 5 included it was intended to be the catch all for everything that doesn’t adhere to QL4.
> 
> I explicitly wanted a level that we could use in a future central quality repository that meant “this package is actually legally or otherwise dangerous to use”.

I also agree it may be good to have a catch all - "not safe to be used in most circumstances"